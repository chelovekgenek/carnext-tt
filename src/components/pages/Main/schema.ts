import * as Yup from "yup"
import { PHONE_MASK, ALPHA_MASK, PromotionCodes } from "helpers"

export const schema = Yup.object().shape({
  firstname: Yup.string()
    .min(2)
    .max(32)
    .matches(ALPHA_MASK)
    .required(),
  surname: Yup.string()
    .min(2)
    .max(32)
    .matches(ALPHA_MASK)
    .required(),
  email: Yup.string()
    .email()
    .required(),
  phone: Yup.string()
    .matches(PHONE_MASK)
    .required(),
  code: Yup.string()
    .test("promotion-code", "promotion code is not valid", value => PromotionCodes.isValid(value))
    .required(),
  agreement: Yup.bool()
    .oneOf([true])
    .required(),
})

export const initialValues: Yup.InferType<typeof schema> = {
  firstname: "",
  surname: "",
  email: "",
  phone: "",
  code: "",
  agreement: true,
}
