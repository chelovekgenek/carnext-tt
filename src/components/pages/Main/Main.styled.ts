import styled, { css } from "styled-components"

export const Row = styled.div`
  display: flex;
`

interface IColumn {
  padding?: number
}
export const Column = styled.div<IColumn>`
  flex-direction: column;
  width: 50%;
  ${({ padding = 2 }) => css`
    &:first-child {
      padding-right: ${padding}rem;
    }
    &:last-child {
      padding-left: ${padding}rem;
    }
  `}
`
