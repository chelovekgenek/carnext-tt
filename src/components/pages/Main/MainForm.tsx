import React from "react"
import { connect } from "react-redux"
import { FormikProps, Form, withFormik } from "formik"

import { Input, Checkbox } from "components/common/form"
import { Button } from "components/common/ui"

import { initialValues, schema } from "./schema"
import * as Styles from "./Main.styled"
import { TAppState } from "store/entities"
import { PromotionActions } from "store/entities/promotion"

interface IDispatchProps {
  submit: typeof PromotionActions.request
}
interface IProps extends IDispatchProps {}

export const MainForm: React.FC<IProps & FormikProps<typeof initialValues>> = ({ handleSubmit, isValid }) => (
  <Form>
    <Styles.Row>
      <Styles.Column padding={1}>
        <Input label="First name" name="firstname" />
      </Styles.Column>
      <Styles.Column padding={1}>
        <Input label="Surname" name="surname" />
      </Styles.Column>
    </Styles.Row>
    <Input label="Email address" name="email" />
    <Input label="Phone number" name="phone" />
    <Input label="Voucher code" name="code" />
    <Checkbox label="Accept Privacy Settlement" name="agreement" />
    <Button type="submit" onClick={handleSubmit} disabled={!isValid}>
      Get offer
      <i className="fas fa-chevron-right" />
    </Button>
  </Form>
)

export default connect<{}, IDispatchProps, {}, TAppState>(
  null,
  {
    submit: PromotionActions.request,
  },
)(
  withFormik<IProps, typeof initialValues>({
    handleSubmit: (values, { props }) => props.submit(values),
    mapPropsToValues: () => initialValues,
    validationSchema: schema,
    validateOnBlur: true,
  })(MainForm),
)
