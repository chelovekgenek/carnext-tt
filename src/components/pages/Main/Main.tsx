import React from "react"

import MainForm from "./MainForm"
import * as Styles from "./Main.styled"

export const Main = () => (
  <Styles.Row>
    <Styles.Column>
      <h1>Get a free year of driving!</h1>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
        consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
        pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est
        laborum.
      </p>
    </Styles.Column>
    <Styles.Column>
      <h1>Get a car leasing offer!</h1>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore
        magna aliqua.
      </p>
      <MainForm />
    </Styles.Column>
  </Styles.Row>
)
