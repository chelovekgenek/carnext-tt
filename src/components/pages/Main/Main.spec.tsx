import React from "react"
import { shallow } from "enzyme"

import { Main } from "./Main"

describe("Page.Main", () => {
  it("should render correctly", () => {
    expect(shallow(<Main />)).toMatchSnapshot()
  })
})
