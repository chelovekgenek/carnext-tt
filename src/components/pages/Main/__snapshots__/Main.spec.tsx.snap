// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page.Main should render correctly 1`] = `
<styled.div>
  <styled.div>
    <h1>
      Get a free year of driving!
    </h1>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
    </p>
  </styled.div>
  <styled.div>
    <h1>
      Get a car leasing offer!
    </h1>
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
    </p>
    <WithFormik(MainForm) />
  </styled.div>
</styled.div>
`;
